src/main/scala/org/allenai/plugins/DeployPlugin.scala

```diff
@@ -459,92 +486,306 @@ object DeployPlugin extends AutoPlugin {
+        )
+      case _: ConfigException.WrongType =>
+        throw new IllegalArgumentException(
+          s"Error: '$key' must be a string in ${config.root().render()}"
+        )
+    }
+  }
 
-    val startArgs = try {
-      hostConfig.getStringList("startup_args").asScala
+  /** Helper method to get replica config overrides from config. */
+  def getConfigOverrides(config: Config): Config = {
+    try {
+      config.getConfig("config_overrides")
     } catch {
-      case _: ConfigException.Missing => Seq()
+      case _: ConfigException.Missing => ConfigFactory.empty
       case _: ConfigException.WrongType =>
         throw new IllegalArgumentException(
-          s"Error: 'startup_args' must be an array of strings in ${hostConfig.root().render()}"
+          s"Error: 'config_overrides' must be an object in ${config.root().render()}"
         )
     }
+  }
+
+  /** Use `find -exec` to locate, stop, and remove stale replicas that won't be updated as part
+    * of a running deploy. Replicas are considered 'stale' if they are running in a directory
+    * that will not be updated by subsequent `rsync` commands.
+    * Directories are chosen to be `rsync`-ed according to the following scheme:
+    *   - If only one replica is being deployed to a host, then the `rsync`-ed directory will be
+    *     `baseDeployDirectory`.
+    *   - If `n > 1` replicas are being deployed to a host, then the `rsync`-ed directories will
+    *     be `baseDeployDirectory-i` for `i` in `1` to `n`.
+    * @param host the remote host to clean up
+    * @param baseDeployDirectory the base deploy directory for this project on the remote host
+    * @param numReplicas the number of replicas that will later be deployed to the remote host
+    *                    using `rsync`
+    * @param sshCommand a constructed SSH command that should be used to communicate with the
+    *                   remote host
+    */
+  def stopStaleReplicas(
+    host: String,
+    baseDeployDirectory: String,
+    numReplicas: Int,
+    sshCommand: Seq[String],
+    log: DeployLogger
+  ): Unit = {
+    // The base deploy directory for this project includes the project name; extract it.
+    val (deployParent, deployTarget) =
+      baseDeployDirectory.splitAt(baseDeployDirectory.lastIndexOf('/'))
+
+    // Strip leading '/' from project name so we can match names on it.
+    val projectName = deployTarget.tail
+
+    // Build a pattern matching the replica directories that we want to maintain for subsequent
+    // updates. This pattern will be passed to `find` as a negated name match.
+    val persistedPattern = if (numReplicas == 1) {
+      // If only one replica is going to be deployed, we want to keep the deploy directory
+      // without a numeric suffix.
+      projectName
+    } else {
+      // If multiple replicas are going to be deployed, we want to keep one deploy directory for
+      // each of them.
+      s"$projectName-[1-$numReplicas]"
```

Comment:
I'm having trouble thinking of a sane solution to this.  Would it be acceptable to just stop every replica matching the pattern `projectName*`, and not delete any directories?